File: csharp/src/code-smells/dispensables/DuplicatedCode.cs
namespace RefactoringAvanzado.CodeSmells.Dispensables;

public static class DuplicatedCode
{
    public static double CalculateOrderTotalWithTax(List<(double Price, int Qty)> items, double taxRate)
    {
        var subtotal = 0.0;
        foreach (var item in items)
        {
            subtotal += item.Price * item.Qty;
        }
        var tax = subtotal * taxRate;
        return subtotal + tax;
    }

    public static double ComputeCartTotalIncludingTax(List<(double Price, int Quantity)> items, double taxRate)
    {
        var subtotal = 0.0;
        foreach (var item in items)
        {
            subtotal += item.Price * item.Quantity;
        }
        var tax = subtotal * taxRate;
        return subtotal + tax;
    }

    public static (double, double) DemoDuplicatedCode()
    {
        var itemsA = new List<(double Price, int Qty)>
        {
            (10, 2),
            (5, 3)
        };
        var itemsB = new List<(double Price, int Quantity)>
        {
            (10, 2),
            (5, 3)
        };
        return (CalculateOrderTotalWithTax(itemsA, 0.21), ComputeCartTotalIncludingTax(itemsB, 0.21));
    }
}

File: typescript/src/code-smells/dispensables/duplicated-code.ts
export function calculateOrderTotalWithTax(
  items: { price: number; qty: number }[],
  taxRate: number,
): number {
  let subtotal = 0
  for (const item of items) {
    subtotal += item.price * item.qty
  }
  const tax = subtotal * taxRate
  return subtotal + tax
}

export function computeCartTotalIncludingTax(
  items: { price: number; quantity: number }[],
  taxRate: number,
): number {
  let subtotal = 0
  for (const item of items) {
    subtotal += item.price * item.quantity
  }
  const tax = subtotal * taxRate
  return subtotal + tax
}

export function demoDuplicatedCode(): [number, number] {
  const itemsA = [
    {price: 10, qty: 2},
    {price: 5, qty: 3},
  ]
  const itemsB = [
    {price: 10, quantity: 2},
    {price: 5, quantity: 3},
  ]
  return [calculateOrderTotalWithTax(itemsA, 0.21), computeCartTotalIncludingTax(itemsB, 0.21)]
}
